src/analysis/signature_help.ml

```diff
@@ -0,0 +1,126 @@
+open Std
+
+open Browse_raw
+
+open Extend_protocol.Reader
+
+type application_signature =
+  { fun_name : string option
+  ; signature : string
+  ; param_offset : (int * int) option
+  }
+
+(* Type variables shared across arguments should all be
+   printed with the same name.
+   [Printtyp.type_scheme] ensure that a name is unique within a given
+   type, but not across different invocations.
+   [reset] followed by calls to [mark_loops] and [type_sch] provide
+   that *)
+let print_type ty env =
+  let module Printtyp = Type_utils.Printtyp in
+  let ppf, to_string = Format.to_string () in
+  Printtyp.wrap_printing_env env ~verbosity:0 (fun () ->
+      Printtyp.mark_loops ty;
+      Printtyp.type_sch ppf ty;
+      to_string ())
+
+(* extract a properly properly parenthesized identifier
+   from (expression_desc (Texp_ident (Longident))) *)
+let extract_ident (exp_desc : Typedtree.expression_desc) =
+  let rec longident ppf : Longident.t -> unit = function
+    | Lident s -> fprintf ppf "%s" (Completion.parenthesize_name s)
+    | Ldot (p, s) -> fprintf ppf "%a.%s" longident p (Completion.parenthesize_name s)
+    | Lapply (p1, p2) -> fprintf ppf "%a(%a)" longident p1 longident p2
+  in
+  match exp_desc with
+  | Texp_ident (_, { txt = li; _ }, _) ->
+    let ppf, to_string = Format.to_string () in
+    longident ppf li;
+    Some (to_string ())
+  | _ -> None
+
+let application_signature path =
+  let module Printtyp = Type_utils.Printtyp in
+  match path with
+  (* provide signature information for applied functions *)
+  | (_, Expression earg)
+    :: (_, Expression { exp_desc =
+      Texp_apply ({ exp_type = { desc = Tarrow _; _ }; _ } as efun, eargs); _ }) :: _ ->
+    let buffer = Buffer.create 16 in
+
+    Printtyp.reset ();
+    let print_type ty = Buffer.add_string buffer (print_type ty efun.exp_env) in
+
+    (* surround function types in parentheses *)
+    let print_parameter_type = function
+      | { Types.desc = Tarrow _; _ } as ty ->
+        Buffer.add_char buffer '(';
```

Comment:
I think you can use `Format.formatter_of_buffer` here. Should be nicer to get rid of all of these intermediate allocations this way. It's not a requirement though

Reply:
If I'm understanding this correctly, using the buffer as a format would require a flush right after or else `Buffer.length` would have incorrect information.

> `formatter_of_buffer b` returns a new formatter writing to buffer `b`. At the end of pretty-printing, the formatter must be flushed using` pp_print_flush` or `pp_print_newline`, to print all the pending material into the buffer.

I think it's cleaner to leave it as it is.

Reply:
Yes, it's certainly uglier. But there's two advantages:
* Proper wrapping for long signatures
* Better perf.

No need to address this now though.